File: sketch/src/main/java/me/panpf/sketch/zoom/ScrollBarHelper.java
/*
 * Copyright (C) 2016 Peng fei Pan <sky@panpf.me>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.panpf.sketch.zoom;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.os.Handler;
import android.os.Looper;
import android.view.animation.DecelerateInterpolator;
import android.widget.ImageView;
import android.widget.Scroller;

import me.panpf.sketch.SLog;
import me.panpf.sketch.util.SketchUtils;

class ScrollBarHelper {

    private ImageZoomer imageZoomer;
    private Paint scrollBarPaint;

    private int scrollBarSize;
    private int scrollBarMargin;
    private int scrollBarRadius;
    private int scrollBarAlpha = 51;
    private RectF scrollBarRectF = new RectF();
    private RectF tempDisplayRectF = new RectF();

    private Handler handler;
    private HiddenScrollBarRunner hiddenScrollBarRunner;
    private FadeScrollBarRunner fadeScrollBarRunner;

    ScrollBarHelper(Context context, ImageZoomer imageZoomer) {
        this.imageZoomer = imageZoomer;

        scrollBarPaint = new Paint();
        scrollBarPaint.setColor(Color.parseColor("#000000"));
        scrollBarPaint.setAlpha(scrollBarAlpha);
        scrollBarSize = SketchUtils.dp2px(context, 3);
        scrollBarMargin = SketchUtils.dp2px(context, 3);
        scrollBarRadius = Math.round(scrollBarSize / 2);

        handler = new Handler(Looper.getMainLooper());
        hiddenScrollBarRunner = new HiddenScrollBarRunner();
        fadeScrollBarRunner = new FadeScrollBarRunner(context);
    }

    void onDraw(Canvas canvas) {
        final RectF drawRectF = tempDisplayRectF;
        imageZoomer.getDrawRect(drawRectF);
        if (drawRectF.isEmpty()) {
            if (SLog.isLoggable(SLog.LEVEL_DEBUG | SLog.TYPE_ZOOM)) {
                SLog.d(ImageZoomer.NAME, "displayRectF is empty. drawScrollBar. drawRectF=%s", drawRectF.toString());
            }
            return;
        }

        Size viewSize = imageZoomer.getViewSize();
        final int viewWidth = viewSize.getWidth();
        final int viewHeight = viewSize.getHeight();
        final float displayWidth = drawRectF.width();
        final float displayHeight = drawRectF.height();

        if (viewWidth <= 0 || viewHeight <= 0 || displayWidth == 0 || displayHeight == 0) {
            if (SLog.isLoggable(SLog.LEVEL_DEBUG | SLog.TYPE_ZOOM)) {
                SLog.d(ImageZoomer.NAME, "size is 0. drawScrollBar. viewSize=%dx%d, displaySize=%sx%s",
                        viewWidth, viewHeight, displayWidth, displayHeight);
            }
            return;
        }

        ImageView imageView = imageZoomer.getImageView();
        final int finalViewWidth = viewWidth - (scrollBarMargin * 2);
        final int finalViewHeight = viewHeight - (scrollBarMargin * 2);

        if ((int) displayWidth > viewWidth) {
            final float widthScale = (float) viewWidth / displayWidth;
            final int horScrollBarWidth = (int) (finalViewWidth * widthScale);

            RectF horScrollBarRectF = scrollBarRectF;
            horScrollBarRectF.setEmpty();
            horScrollBarRectF.left = imageView.getPaddingLeft() + scrollBarMargin + (drawRectF.left < 0 ? (int) ((Math.abs(drawRectF.left) / drawRectF.width()) * finalViewWidth) : 0);
            horScrollBarRectF.top = imageView.getPaddingTop() + scrollBarMargin + finalViewHeight - scrollBarSize;
            horScrollBarRectF.right = horScrollBarRectF.left + horScrollBarWidth;
            horScrollBarRectF.bottom = horScrollBarRectF.top + scrollBarSize;
            canvas.drawRoundRect(horScrollBarRectF, scrollBarRadius, scrollBarRadius, scrollBarPaint);
        }

        if ((int) displayHeight > viewHeight) {
            final float heightScale = (float) viewHeight / displayHeight;
            final int verScrollBarHeight = (int) (finalViewHeight * heightScale);

            RectF verScrollBarRectF = scrollBarRectF;
            verScrollBarRectF.setEmpty();
            verScrollBarRectF.left = imageView.getPaddingLeft() + scrollBarMargin + finalViewWidth - scrollBarSize;
            verScrollBarRectF.top = imageView.getPaddingTop() + scrollBarMargin + (drawRectF.top < 0 ? (int) ((Math.abs(drawRectF.top) / drawRectF.height()) * finalViewHeight) : 0);
            verScrollBarRectF.right = verScrollBarRectF.left + scrollBarSize;
            verScrollBarRectF.bottom = verScrollBarRectF.top + verScrollBarHeight;
            canvas.drawRoundRect(verScrollBarRectF, scrollBarRadius, scrollBarRadius, scrollBarPaint);
        }
    }

    /**
     * 此方法里没有执行 imageView.invalidate()，因为回调的地方会有执行
     */
    void onMatrixChanged() {
        scrollBarPaint.setAlpha(scrollBarAlpha);

        if (fadeScrollBarRunner.isRunning()) {
            fadeScrollBarRunner.abort();
        }
        handler.removeCallbacks(hiddenScrollBarRunner);

        handler.postDelayed(hiddenScrollBarRunner, 800);
    }

    private void invalidateView() {
        ImageView imageView = imageZoomer.getImageView();
        if (imageView != null) {
            imageView.invalidate();
        }
    }

    private class HiddenScrollBarRunner implements Runnable {

        @Override
        public void run() {
            fadeScrollBarRunner.start();
        }
    }

    private class FadeScrollBarRunner implements Runnable {
        private Scroller scroller;

        FadeScrollBarRunner(Context context) {
            this.scroller = new Scroller(context, new DecelerateInterpolator());
            scroller.forceFinished(true);
        }

        public void start() {
            scroller.startScroll(scrollBarAlpha, 0, -scrollBarAlpha, 0, 300);
            handler.post(this);
        }

        boolean isRunning() {
            return !scroller.isFinished();
        }

        void abort() {
            scroller.forceFinished(true);
        }

        @Override
        public void run() {
            if (scroller.computeScrollOffset()) {
                scrollBarPaint.setAlpha(scroller.getCurrX());
                invalidateView();
                handler.postDelayed(this, 60);
            }
        }
    }
}

File: sketch-zoom/src/main/java/com/github/panpf/sketch/zoom/internal/ScrollBarHelper.kt
/*
 * Copyright (C) 2019 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.zoom.internal

import android.content.Context
import android.content.res.Resources
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.view.animation.DecelerateInterpolator
import android.widget.Scroller
import androidx.core.view.ViewCompat
import com.github.panpf.sketch.zoom.Zoomer
import kotlin.math.abs
import kotlin.math.roundToInt

internal class ScrollBarHelper(context: Context, private val zoomer: Zoomer) {

    private val scrollBarSize: Float = 3f * Resources.getSystem().displayMetrics.density
    private val scrollBarMargin: Float = 3f * Resources.getSystem().displayMetrics.density
    private val scrollBarRadius: Int = (scrollBarSize / 2).roundToInt()
    private val scrollBarAlpha: Int = 51
    private val scrollBarPaint: Paint = Paint().apply {
        color = Color.parseColor("#000000")
        alpha = scrollBarAlpha
    }
    private val view = zoomer.view
    private val drawRectF = RectF()
    private val scrollBarRectF = RectF()
    private val fadeRunnable: FadeRunnable = FadeRunnable(context, this)
    private val delayFadeRunnable: DelayFadeRunnable = DelayFadeRunnable(this, fadeRunnable)

    fun onDraw(canvas: Canvas) {
        val drawRectF = drawRectF.apply {
            zoomer.getDrawRect(this)
        }.takeIf { !it.isEmpty } ?: return
        val (viewWidth, viewHeight) = zoomer.viewSize.takeIf { !it.isEmpty } ?: return
        val drawWidth = drawRectF.width()
        val drawHeight = drawRectF.height()
        val viewAvailableWidth =
            viewWidth - (scrollBarMargin * 2) - view.paddingLeft - view.paddingRight
        val viewAvailableHeight =
            viewHeight - (scrollBarMargin * 2) - view.paddingTop - view.paddingBottom

        // draw hor scroll bar
        if (drawWidth.toInt() > viewWidth) {
            val widthScale = viewWidth.toFloat() / drawWidth
            val horScrollBarWidth =
                (viewAvailableWidth * widthScale).coerceAtLeast(scrollBarSize).toInt()
            val horScrollBarRectF = scrollBarRectF.apply {
                val mapLeft = if (drawRectF.left < 0) {
                    (abs(drawRectF.left) / drawRectF.width() * viewAvailableWidth).toInt()
                } else 0
                left = (view.paddingLeft + scrollBarMargin + mapLeft)
                right = left + horScrollBarWidth
                top = (view.paddingTop + scrollBarMargin + viewAvailableHeight - scrollBarSize)
                bottom = top + scrollBarSize
            }
            canvas.drawRoundRect(
                horScrollBarRectF,
                scrollBarRadius.toFloat(),
                scrollBarRadius.toFloat(),
                scrollBarPaint
            )
        }

        // draw ver scroll bar
        if (drawHeight.toInt() > viewHeight) {
            val heightScale = viewHeight.toFloat() / drawHeight
            val verScrollBarHeight =
                (viewAvailableHeight * heightScale).coerceAtLeast(scrollBarSize).toInt()
            val verScrollBarRectF = scrollBarRectF.apply {
                val mapTop = if (drawRectF.top < 0) {
                    (abs(drawRectF.top) / drawRectF.height() * viewAvailableHeight).toInt()
                } else 0
                left = (view.paddingLeft + scrollBarMargin + viewAvailableWidth - scrollBarSize)
                right = left + scrollBarSize
                top = (view.paddingTop + scrollBarMargin + mapTop)
                bottom = top + verScrollBarHeight
            }
            canvas.drawRoundRect(
                verScrollBarRectF,
                scrollBarRadius.toFloat(),
                scrollBarRadius.toFloat(),
                scrollBarPaint
            )
        }
    }

    fun onMatrixChanged() {
        scrollBarPaint.alpha = scrollBarAlpha
        if (fadeRunnable.isRunning) {
            fadeRunnable.cancel()
        }
        delayFadeRunnable.start()
    }

    private class DelayFadeRunnable(
        val scrollBarHelper: ScrollBarHelper,
        val fadeRunnable: FadeRunnable
    ) : Runnable {

        override fun run() {
            fadeRunnable.start()
        }

        fun start() {
            scrollBarHelper.view.removeCallbacks(this)
            scrollBarHelper.view.postDelayed(this, 800)
        }
    }

    private class FadeRunnable(context: Context, val scrollBarHelper: ScrollBarHelper) : Runnable {

        private val scroller: Scroller = Scroller(context, DecelerateInterpolator())

        val isRunning: Boolean
            get() = !scroller.isFinished

        fun start() {
            cancel()

            val startX = scrollBarHelper.scrollBarAlpha
            val dx = -startX
            scroller.startScroll(startX, 0, dx, 0, 300)
            scrollBarHelper.view.post(this)
        }

        fun cancel() {
            scrollBarHelper.view.removeCallbacks(this)
            scroller.forceFinished(true)
        }

        override fun run() {
            if (!scroller.isFinished && scroller.computeScrollOffset()) {
                scrollBarHelper.scrollBarPaint.alpha = scroller.currX
                scrollBarHelper.view.invalidate()
                ViewCompat.postOnAnimation(scrollBarHelper.view, this)
            }
        }
    }
}